WIP: Basic UAVCAN integration

antoinealb opened the pull request:
This PR adds a basic UAVCAN node. For now it only publishes its node status on the bus using a fixed ID (42). I had to use some tricks in order to make libuavcan fit into this ressource constrained system:

* Implement some C++ stubs/syscalls to reduce flash footprint
* Basic C++ compiler tweaking such as disabling exceptions and RTTI
* Run the UAVCAN node in the main thread to avoid having a separate stack consuming RAM
* Disabling threadsafe statics to reduce flash footprint
* Using smaller buffers for UAVCAN queues (might be an issue, needs to be tested)
* Disabling UAVCAN assertions. This was the biggest gain (about 30kB of flash), but it suppresses some useful checks. In the future I would like to have a minimalistic handling of those assertions but for now it sounds good enough. Note that the assertions are still used when we don't use the bootloader so we can still do some basic sanity checking on hardware.

I still need to test that on hardware.

@nuft what do you think ?

antoinealb committed:
Use rules.ld and rules.mk from ChibiOS

antoinealb committed:
Remove useless functions

antoinealb committed:
format

antoinealb committed:
Use CVRA package manager

antoinealb committed:
Add UAVCAN to build

antoinealb committed:
Add custom syscall implementation to reduce size

antoinealb committed:
Disable exceptions

This allows us to shave an additional few kb

antoinealb committed:
Create a UAVCAN node

To save up some RAM it is spawned from the main thread.

antoinealb committed:
Disable threadsafe statics

Won't be needed and saves some code

antoinealb committed:
Disable assertions when building with bootloader

This gives us a lot of headroom (~30k).

antoinealb committed:
Add dependencies as submodules

msplr commented on Makefile:
```diff
@@ -194,9 +220,15 @@ ULIBS =
 ##############################################################################
 
-RULESPATH = .
+RULESPATH = $(CHIBIOS)/os/common/ports/ARMCMX/compilers/gcc/
```
typo here. should be `RULESPATH = $(CHIBIOS)/os/common/ports/ARMCMx/compilers/GCC`

antoinealb replied:
Correct, I will fix it.

msplr commented:
good job 👍 

I'm worried that the binary size will augment later on, because now we don't use any UAVCAN messages - to be verified.

antoinealb commented:
Ok, I changed the rules path (`GCC`) and rebased it in my commits. Can you merge it ?

Also for the binary size so far we are using the node status message (published periodically), so I think we should be fine. Should the size augment too much we will always find a way to make it fit.

antoinealb committed:
Read node configuration from bootloader
